### PR TITLE
Pass also precursor m/z to the similarity functions and add GNPS score

### PR DESCRIPTION
**Pull request by jorainer**
This PR adds the following changes:
- pass also precursor m/z values of the two spectra that are compared in `compareSpectra` to the spectra comparison function `FUN` (issue #171).
- add `joinPeaksGnps` and `gnps` functions to perform a peak matching and similarity score calculation as done by GNPS (issue #171).

### CODE REVIEW

File: R/peaks-functions.R [hunk 1]
```diff
@@ -224,6 +224,20 @@ NULL
 #'   between `x` and `y` are returned by `type = "inner"`, i.e. only
 #'   peaks present in both spectra are reported.
 #'
+#' - `joinPeaksGnps` matches/maps peaks between spectra with an approach
+#'   similar to the one used in GNPS: peaks are considered matching if a) the
```
Comment (lgatto) [inline R/peaks-functions.R:228]: Could you clarify what you mean by *similar*? Same algorithm but different implementation? Any tests that assess if identical or what the deviations are?

File: R/peaks-functions.R [hunk 2]
```diff
@@ -286,11 +310,150 @@ NULL
...
+#'     first peak (row) of `y` or `NA_real_` if the peak does not match.
+#'
+#' @param y `numeric` peak `matrix` with m/z (first column) and intensity
+#'     values (second column). Rows in `x` and `y` are considered to contain
+#'     matching peaks, i.e. the first row in `x` contains a peak matching the
+#'     first peak (row) of `y` or `NA_real_` if the peak does not match.
+#'
+#' @param ... ignored
+#'
+#' @author Johannes Rainer, Michael Witting, based on the code from the ref
```
Comment (lgatto) [inline R/peaks-functions.R:396]: ... based on code from Xing et al. (2020).

File: R/peaks-functions.R [hunk 3]
```diff
@@ -286,11 +310,150 @@ NULL
...
+#' pmz_x <- 91
+#' pmz_y <- 105
+#'
+#' ## joinPeaksGnps finds 4 matches
+#' map <- joinPeaksGnps(x, y, pmz_x, pmz_y)
+#' map
+#'
+#' ## Calculate similarity
+#' gnps(map$x, map$y)
+gnps <- function(x, y, ...) {
```
Comment (lgatto) [inline R/peaks-functions.R:426]: Should this function not rather be added to `MsCoreUtils`? It doesn't use or return `Spectra` objects.

File: R/peaks-functions.R (hunk 1 above)
Comment (jorainer) [reply R/peaks-functions.R:228]: It's actually the same approach but different implementation (GNPS is based on python). There are differences, but after the 3rd decimal.

File: R/peaks-functions.R (hunk 2 above)
Comment (jorainer) [reply R/peaks-functions.R:396]: I will update.

File: R/peaks-functions.R (hunk 1 above)
Comment (lgatto) [reply R/peaks-functions.R:228]: May be simply mention this in the documentation.

File: R/peaks-functions.R (hunk 3 above)
Comment (jorainer) [reply R/peaks-functions.R:426]: I originally also thought the same. But this function needs to be called in combination with `joinPeaksGnps` to get GNPS scores. Putting it into `MsCoreUtils` I'm afraid people might get the impression that this alone calculates GNPS similarity. But I have no strong opinion on that. I could also add it to `MsCoreUtils` if you think that's better.

File: R/peaks-functions.R (hunk 2 above)
Comment (jorainer) [reply R/peaks-functions.R]: Done

File: R/peaks-functions.R (hunk 1 above)
Comment (jorainer) [reply R/peaks-functions.R]: Done

File: R/peaks-functions.R (hunk 3 above)
Comment (lgatto) [reply R/peaks-functions.R:427]: Does that mean that the function on its own is likely to be useless? If so, them I would suggest to keep it in `Spectra` but not export it. If there is a use for it, then we a move to `MsCoreUtils` could be considered.

Comment (jorainer) [reply R/peaks-functions.R:427]: Maybe I should explain  better: in order to calculate GNPS scores one would have to call:

```r
compareSpectra(sps, sps2, MAPFUN = joinPeaksGnps, FUN = gnps)
```

so we need to export `gnps`.  I could move `gnps` to `MsCoreUtils`, but then I would need also to add a function `join_gnps` there (that does the peak matching as in GNPS) and mention in the documentation that `gnps` only returns GNPS scores, if the peaks were first matched with `join_gnps`.

Comment (jorainer) [reply R/peaks-functions.R:427]: OK, in the end I agree. I will add the `join_gnps` and `gnps` functions to the `MsCoreUtils` package and then update the PR here. So you will get another review request today over in `MsCoreUtils` :)

**Pull request closed without merging**